File: packages/service-auth-py/src/fuzefront_service_auth/verifier.py
"""Callee-side verification of FuzeFront M2M bearer tokens.

Contract: `POST /api/v1/security/tokens/introspect` in
`packages/security/openapi.yaml` (`TokenIntrospectRequest` ->
`TokenIntrospection`). The server implementation
(`backend/security/src/routes/security.ts`) ALWAYS answers this endpoint
with HTTP 200 -- fail-closed introspection is expressed entirely in the
response BODY's `active: boolean` field, never in the status code:

    router.post('/tokens/introspect', async (req, res) => {
      try {
        ...
        res.status(200).json(r)
      } catch (err) {
        // Fail-closed: introspection never throws to the caller.
        res.status(200).json({ active: false })
      }
    })

This is the single most important behaviour in this package. Branching on
HTTP status ("2xx means good") fails OPEN the moment a caller does not read
the body -- see `test_verifier.py::test_inactive_token_with_http_200_is_rejected`
for the regression test this package exists to guarantee.
"""

from __future__ import annotations

import json
import time
from dataclasses import dataclass, field
from typing import Any, Dict, Optional

from ._cache import DEFAULT_MAX_SIZE, DEFAULT_MAX_TTL_SECONDS, PositiveCache
from ._http import HttpPost, default_http_post
from .exceptions import TokenVerificationError

DEFAULT_TIMEOUT_SECONDS = 10.0


@dataclass(frozen=True)
class MachineIdentity:
    """A verified machine caller, projected from `TokenIntrospection`."""

    subject: str
    scope: Optional[str] = None
    tenant_id: Optional[str] = None
    expires_at: Optional[int] = None
    raw: Dict[str, Any] = field(default_factory=dict)


class MachineTokenVerifier:
    """Verifies a bearer token against FuzeFront's `/tokens/introspect` endpoint.

    Fail-closed on EVERY ambiguity -- a connection error, a timeout, a
    malformed body, a body with `active` missing or non-boolean, or
    `active: false` -- all raise `TokenVerificationError`. There is no code
    path in this class that treats an error as "allow".

    Positive results (only) are cached in memory, bounded by size and by the
    token's own `exp` claim, so a hot path does not re-introspect on every
    request while still respecting revocation within one cache TTL window.

    Example:
        verifier = MachineTokenVerifier(base_url="http://fuzefront-security-service:3000")
        identity = verifier.verify_machine_token(bearer_token)  # raises on any failure
    """

    def __init__(
        self,
        base_url: str,
        *,
        timeout: float = DEFAULT_TIMEOUT_SECONDS,
        http_post: Optional[HttpPost] = None,
        cache_max_size: int = DEFAULT_MAX_SIZE,
        cache_max_ttl_seconds: float = DEFAULT_MAX_TTL_SECONDS,
        clock=time.time,
    ) -> None:
        if not base_url:
            raise ValueError("base_url is required")
        self._base_url = base_url.rstrip("/")
        self._timeout = timeout
        self._http_post = http_post or default_http_post
        self._clock = clock
        self._cache: PositiveCache[MachineIdentity] = PositiveCache(
            max_size=cache_max_size, max_ttl_seconds=cache_max_ttl_seconds
        )

    def verify_machine_token(self, token: str) -> MachineIdentity:
        """Return the caller's `MachineIdentity`, or raise `TokenVerificationError`.

        Fail-closed: any ambiguity (network failure, timeout, malformed
        body, missing/non-bool `active`, or `active: false`) raises. Never
        returns a value for a token that is not affirmatively active.
        """
        if not token or not isinstance(token, str):
            raise TokenVerificationError("no bearer token presented")

        now = self._clock()
        cached = self._cache.get(token, now)
        if cached is not None:
            return cached

        try:
            status, body = self._http_post(
                f"{self._base_url}/api/v1/security/tokens/introspect",
                {"token": token},
                self._timeout,
            )
        except Exception as error:  # noqa: BLE001 - fail closed on ANY transport error
            raise TokenVerificationError(
                f"introspection request failed (failing closed): {error}"
            ) from error

        # The contract guarantees HTTP 200 always. We still fail closed if a
        # proxy/misconfiguration ever produces a non-200 -- we do NOT treat
        # "not 200" as "must therefore be inactive and safe to say so
        # positively"; we simply refuse to authenticate the caller.
        if status != 200:
            raise TokenVerificationError(
                f"introspection endpoint returned unexpected HTTP {status} (failing closed)"
            )

        try:
            data = json.loads(body)
        except (TypeError, ValueError) as error:
            raise TokenVerificationError(
                f"malformed introspection response body (failing closed): {error}"
            ) from error

        if not isinstance(data, dict):
            raise TokenVerificationError("introspection response was not a JSON object (failing closed)")

        active = data.get("active")
        if not isinstance(active, bool):
            raise TokenVerificationError(
                "introspection response missing boolean 'active' (failing closed)"
            )

        # THE branch this package exists for: HTTP 200 tells us nothing.
        # Only `active is True` authenticates the caller.
        if active is not True:
            raise TokenVerificationError("token is not active")

        subject = data.get("subject")
        if not isinstance(subject, str) or not subject:
            raise TokenVerificationError(
                "introspection response marked active but is missing 'subject' (failing closed)"
            )

        identity = MachineIdentity(
            subject=subject,
            scope=data.get("scope"),
            tenant_id=data.get("tenantId"),
            expires_at=data.get("expiresAt"),
            raw=data,
        )
        self._cache.put(token, identity, now, expires_at=data.get("expiresAt"))
        return identity

File: packages/service-auth/src/verifier.ts
/**
 * @fuzefront/service-auth — the RESOURCE SERVER half.
 *
 * `POST /api/v1/security/tokens/introspect` ALWAYS answers HTTP 200 — per
 * `packages/security/openapi.yaml`'s `TokenIntrospection` schema, the wire
 * contract is "fail-closed IN THE BODY": an unknown/expired/revoked token
 * still comes back 200 with `{ active: false }`. A caller that branches on
 * status code instead of the `active` boolean fails OPEN — every 200 would
 * look like success. This module branches on `active`, never on status, and
 * that is the entire reason it exists rather than "just call fetch".
 *
 * Every ambiguity is a denial: network error, timeout, non-200 status,
 * unparsable body, a body missing `active`, `active` typed as anything but a
 * boolean. None of those return a permissive identity — they throw.
 */

import { FetchLike, MachineIdentity, ServiceAuthError, TokenIntrospection } from './types';

const INTROSPECT_PATH = '/api/v1/security/tokens/introspect';
const DEFAULT_TIMEOUT_MS = 3000;
const DEFAULT_CACHE_TTL_SECONDS = 5;
const DEFAULT_CACHE_MAX_ENTRIES = 1000;

export interface MachineTokenVerifierOptions {
  /** FuzeFront's same-origin API base, e.g. `https://app.fuzefront.com/api`. */
  baseUrl: string;
  /** Inject an alternative fetch (tests, old Node). Defaults to `globalThis.fetch`. */
  fetch?: FetchLike;
  /** Per-request timeout. Default 3000ms. */
  timeoutMs?: number;
  /**
   * How long a POSITIVE (`active: true`) result may be reused for the same
   * token, in seconds. Default 5. Set to 0 to disable caching entirely.
   *
   * NEGATIVE results are NEVER cached, at any setting — an inactive/expired/
   * revoked verdict is always re-asked next time, so a revocation is visible
   * on the very next call rather than being masked by a stale cache entry.
   * The cache also never outlives the token's own `exp`.
   */
  cacheTtlSeconds?: number;
  /** Bound on cache size (FIFO eviction). Default 1000. */
  cacheMaxEntries?: number;
}

export interface MachineTokenVerifier {
  /**
   * Introspect `token` and return its normalized identity, or throw
   * `ServiceAuthError`. NEVER returns a value for an inactive/undecidable
   * token — every failure mode is a throw.
   */
  verifyMachineToken(token: string): Promise<MachineIdentity>;
}

interface CacheEntry {
  identity: MachineIdentity;
  expiresAtMs: number;
}

/** Build a fail-closed verifier bound to FuzeFront's introspection endpoint. */
export function createMachineTokenVerifier(options: MachineTokenVerifierOptions): MachineTokenVerifier {
  if (!options?.baseUrl) {
    throw new ServiceAuthError(
      'MISCONFIGURED',
      'createMachineTokenVerifier requires a baseUrl pointing at the FuzeFront Security API.',
      500,
    );
  }

  const base = options.baseUrl.replace(/\/+$/, '');
  const timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  const cacheTtlMs = (options.cacheTtlSeconds ?? DEFAULT_CACHE_TTL_SECONDS) * 1000;
  const cacheMaxEntries = options.cacheMaxEntries ?? DEFAULT_CACHE_MAX_ENTRIES;
  // Only ever holds ACTIVE identities. A cache miss means "ask introspection
  // again", never "assume inactive" and never "assume active".
  const cache = new Map<string, CacheEntry>();

  const resolvedFetch: FetchLike | undefined =
    options.fetch ??
    (typeof globalThis.fetch === 'function'
      ? (globalThis.fetch.bind(globalThis) as unknown as FetchLike)
      : undefined);

  if (!resolvedFetch) {
    throw new ServiceAuthError(
      'MISCONFIGURED',
      'No fetch implementation available. Pass `fetch` explicitly on Node <18.',
      500,
    );
  }

  function cachedIdentity(token: string): MachineIdentity | undefined {
    if (cacheTtlMs <= 0) return undefined;
    const hit = cache.get(token);
    if (!hit) return undefined;
    if (hit.expiresAtMs <= Date.now()) {
      cache.delete(token);
      return undefined;
    }
    return hit.identity;
  }

  function remember(token: string, identity: MachineIdentity): void {
    if (cacheTtlMs <= 0) return;
    if (cache.size >= cacheMaxEntries) {
      const oldest = cache.keys().next();
      if (!oldest.done) cache.delete(oldest.value);
    }
    // Never cache past the token's own expiry, on top of the configured TTL.
    let ttl = cacheTtlMs;
    if (typeof identity.expiresAt === 'number') {
      const tokenTtlMs = identity.expiresAt * 1000 - Date.now();
      ttl = Math.max(0, Math.min(ttl, tokenTtlMs));
    }
    if (ttl <= 0) return;
    cache.set(token, { identity, expiresAtMs: Date.now() + ttl });
  }

  async function introspect(token: string): Promise<TokenIntrospection> {
    const controller = typeof AbortController === 'function' ? new AbortController() : undefined;
    const timer = controller ? setTimeout(() => controller.abort(), timeoutMs) : undefined;
    let res: Awaited<ReturnType<FetchLike>>;
    try {
      res = await resolvedFetch!(`${base}${INTROSPECT_PATH}`, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ token }),
        signal: controller?.signal,
      });
    } catch (err) {
      // Network error, DNS failure, connection reset, timeout (abort). Every
      // one of these is undecidable => deny, never "assume active".
      throw new ServiceAuthError(
        'INTROSPECTION_UNAVAILABLE',
        `Introspection request failed: ${(err as Error)?.message ?? 'unknown error'}; denying.`,
        401,
      );
    } finally {
      if (timer) clearTimeout(timer);
    }

    if (!res.ok) {
      // The contract says introspection always answers 200. A non-200 here
      // means something ELSE went wrong in front of it (proxy, LB, outage) —
      // still undecidable, still a denial. This is deliberately NOT treated
      // the same as an in-body `active: false`.
      throw new ServiceAuthError(
        'INTROSPECTION_UNAVAILABLE',
        `Introspection returned unexpected status ${res.status}; denying.`,
        401,
      );
    }

    let body: unknown;
    try {
      body = await res.json();
    } catch (err) {
      throw new ServiceAuthError(
        'MALFORMED_RESPONSE',
        `Introspection returned a non-JSON body: ${(err as Error)?.message ?? 'parse error'}; denying.`,
        401,
      );
    }

    const active = (body as { active?: unknown } | null)?.active;
    if (typeof active !== 'boolean') {
      // Missing/mistyped `active` is exactly as dangerous as a false one
      // treated as true would be. Deny.
      throw new ServiceAuthError(
        'MALFORMED_RESPONSE',
        'Introspection response is missing a boolean `active`; denying.',
        401,
      );
    }

    return body as TokenIntrospection;
  }

  return {
    async verifyMachineToken(token: string): Promise<MachineIdentity> {
      if (!token) {
        throw new ServiceAuthError('NO_TOKEN', 'No token presented to verify.', 401);
      }

      const cached = cachedIdentity(token);
      if (cached) return cached;

      // THIS is the branch the whole package exists to get right: the HTTP
      // call above always resolves with status 200 on success. Whether the
      // token is usable is decided here, from the body — never from `res.ok`.
      const introspection = await introspect(token);

      if (!introspection.active) {
        throw new ServiceAuthError('TOKEN_INACTIVE', 'Token is not active.', 401);
      }
      if (!introspection.subject) {
        throw new ServiceAuthError(
          'MALFORMED_RESPONSE',
          'Active introspection result is missing `subject`; denying.',
          401,
        );
      }

      const identity: MachineIdentity = {
        subject: introspection.subject,
        tenantId: introspection.tenantId ?? null,
        scope: introspection.scope,
        scopes: introspection.scope ? introspection.scope.split(/\s+/).filter(Boolean) : [],
        expiresAt: introspection.expiresAt,
        raw: introspection,
      };

      remember(token, identity);
      return identity;
    },
  };
}
